### PR TITLE
ARROW-14717: [Go] Use the ipc.Reader allocator in messageReader

### PR DESCRIPTION
**Pull request by chriscasola**
Allocate the body of messages from the memory.Allocator instead
of using make.

### CODE REVIEW

File: go/arrow/ipc/message.go
Comment (zeroshade) [inline go/arrow/ipc/message.go:162]: do we really need to have a full options addition here as opposed to just taking the allocator as a param? Not sure if there's any other configurable options we'd want to be adding in the future, particularly if we're going to ignore any options that aren't the allocator here.

Comment (chriscasola) [reply go/arrow/ipc/message.go:162]: I wanted to avoid a breaking change by adding a param. Is that not an issue? Or are you suggesting a new function?

Comment (zeroshade) [reply go/arrow/ipc/message.go:162]: You're right about avoiding breaking with the options. Though I personally doubt many people are calling `NewMessageReader` directly, it's really only exported because the `flight` package uses it to sit on top of the ipc readers to provide access to the additional metadata in flight messages.

It's not a big issue either way I think. More of an aesthetic thing where this gives the impression that message reader would work with other options that get ignored like the reader is doing now.

**Comment by zeroshade** [comment]
You are also going to want to modify https://github.com/apache/arrow/blob/master/go/arrow/ipc/file_reader.go#L350 where the decompression is handled.

**Comment by zeroshade** [comment]
Honestly, looking at this and now that I think of it. It looks like the readers actually call `bytes.NewReader` on the buffer that was already created in the message reader. As a result, it looks like we're currently actually copying no matter what which sucks, but makes it easy to stick the allocator in there. 

Optimally, we'd do what you've done here with the allocator in the message reader for the body, and then just pass slices of this buffer when creating the records. Though that would probably be beyond the scope of this change.

File: go/arrow/ipc/message.go
Comment (chriscasola) [reply go/arrow/ipc/message.go:162]: The use of the options pattern is a bit confusing in this library, especially the `WithSchema` option, since in a lot of cases it's not optional and the function will panic without it. In my opinion required things should be real parameters.

I'm going to leave this as-is unless you feel strongly about making it a normal parameter and are okay with the breaking change.

**Comment by chriscasola** [comment]
@zeroshade I see the issue with `bytes.NewReader` but agree that sounds like a more involved change. Should I still modify the decompression logic in this PR?

**Comment by zeroshade** [comment]
@chriscasola make sure to properly call `.Release` on the new `memory.Buffer`s you're creating since now you're building them off the allocator, and then I'd say it's a good idea to push it down to the decompression logic here too, otherwise it still won't be accurate accounting i think.

**Comment by chriscasola** [comment]
@zeroshade should be good now.

File: go/arrow/ipc/file_reader.go
Comment (zeroshade) [inline go/arrow/ipc/file_reader.go:348]:
```go
for i, field := range schema.Fields() {
        cols[i] = ctx.loadArray(field.Type)
        defer cols[i].Release()
}
```

No need for the separate defer func and loop

File: go/arrow/ipc/message.go
Comment (zeroshade) [reply go/arrow/ipc/message.go:162]: I agree with you on the `WithSchema` comment as far as for the writer (it's not required for the reader which will read the schema from the stream, and you only use `WithSchema` for asserting that you're reading a specific schema). 

For now we'll leave it, but after the release of v6.0.1 which will properly tag this so that people can upgrade to v6, and people start using the tagged releases, then i'll be comfortable making the breaking change since it would be in a tagged *major release* that would make sense to allow a breaking change in.

File: go/arrow/ipc/message.go
Comment (zeroshade) [inline go/arrow/ipc/message.go:235]: the message in `newMessageFromFB` calls retain on the body buffer?

File: go/arrow/ipc/reader.go
Comment (zeroshade) [inline go/arrow/ipc/reader.go:78]: ugh, this feels horrible, but i can't think of a good way to improve this at the moment....

File: go/arrow/memory/checked_allocator.go
Comment (chriscasola) [reply go/arrow/memory/checked_allocator.go:30]: There is no `atomic.LoadInt`

**Comment by zeroshade** [comment]
@chriscasola LGTM for the most part, just a couple of comments i'd like addressed first.

File: go/arrow/ipc/file_reader.go
Comment (chriscasola) [reply go/arrow/ipc/file_reader.go:348]: I didn't want to call `defer` inside a loop - is that ok?

File: go/arrow/ipc/message.go
Comment (chriscasola) [reply go/arrow/ipc/message.go:235]: `newMessageFromFB` calls `Retain` on the body.

File: go/arrow/ipc/file_reader.go
Comment (zeroshade) [reply go/arrow/ipc/file_reader.go:348]: yup, it's perfectly okay. Remember that defer's only run at the end of the enclosing *function* not scope.

**Comment by chriscasola** [comment]
@zeroshade this is ready for another look.

**Comment by zeroshade** [comment]
@chriscasola This LGTM, once all the checks pass I'll merge it.

**Review by zeroshade — APPROVED** [review]
+1

**Pull request closed without merging**
